Docs website

gabro opened the pull request:
Preview it here: http://scalameta.org/metals

olafurpg commented on docs/users/installation.md:
```diff
@@ -0,0 +1,4 @@
+# Sorry, nothing to see here
```
👍 👍 👍

olafurpg commented:
I noticed that the title contains "Gitbook", is there a way to remove that?

gabro commented:
> Is it possible to include a logo ?

Where?

> Is it possible to list the subsections/TOC in the left sidebar only for the open page?

There's a few plugins for TOC, but they include it at the top of the page. I may end up writing a custom plugin.

> I noticed that the title contains "Gitbook", is there a way to remove that?

Done

> can we customize the favicon? I don't know if the scalameta logo looks good as a favicon, any suggestions?

Yep, done.

> would it be possible to hide the sidebar by default?

Why?

olafurpg commented:
> Where? (logo)

Good question 🤔 It is possible to replace the landing page part of the sidebar to have logo + "Metals" ?

<img width="152" alt="overview_ _gitbook" src="https://user-images.githubusercontent.com/1408093/39098813-a5b5a306-4670-11e8-8bba-0fc9dfbfbdfc.png">


> Why? (hide sidebar)

It would maybe be nice to let the landing page be more standalone, the sidebar could be open by default in all other pages. I don't feel strongly about this though

> There's a few plugins for TOC, but they include it at the top of the page. I may end up writing a custom plugin.

Not critical IMO, but for large pages it's nice to have the TOC for the open page

gabro commented:
> Good question 🤔 It is possible to replace the landing page part of the sidebar to have logo + "Metals" ?

![image](https://user-images.githubusercontent.com/691940/39098870-8ded6fb4-4671-11e8-80e3-085730c6d891.png)


This is what I have so far using https://www.npmjs.com/package/gitbook-plugin-insert-logo. I'll work on putting the title on.

olafurpg commented:
Looking at http://scalameta.org/metals/contributors/new-editor/  now with the material theme and I really like it! Readable and easy to find content, which is most important for me.

![screen shot 2018-04-23 at 11 04 02](https://user-images.githubusercontent.com/1408093/39116977-12ace9cc-46e6-11e8-92d6-1250dfb48f91.png)

Some comments

- The markdown rendering for nested bullets is definitely better than in gitbook.
- I love the TOC for the open page, I think that makes it very nice to browse the content especially for large pages (I like large pages 😅 ).
- the current color is quite bland, I would like something more bright and colorful cc/ @mads-hartmann
- The scalameta favicon looks better than I expected 👍 
- The white background for logo is a bit weird, and the shadow on hover is weird too
![screen shot 2018-04-23 at 11 06 23](https://user-images.githubusercontent.com/1408093/39117090-645965ca-46e6-11e8-967f-2a08a8f38f04.png)
- anchor links to headers are difficult to find since they're floated to the right, it would be better to have then next to the header
- wide code listing have horizontal scroll 👍 
- "edit on github" is accessible 👍 
- is it possible to have a splashy landing page like https://docusaurus.io/ or https://bucklescript.github.io/ ? (I'm getting greedy 😏 )
- does hugo have something for search? I like how it looks in https://squidfunk.github.io/mkdocs-material/. Alternatively, we can setup algolia search

PS. I opened a ticket in scalafmt (https://github.com/scalameta/scalafmt/issues/1146) to synchronize the work here with scalafmt.org, I would love to reuse the same setup for scalafmt, scalameta, metals, metadoc, and maybe scalafix too.

olafurpg commented:
No rush, whenever you have time!

mads-hartmann commented:
I think the site looks nice and clean. It's easy to find the content you care about 👍

At Famly we've recently started using mkdocs with their material theme for internal docs for our operations. I don't think it's powerful enough for what you want but their search works amazingly and I really love that you have the site navigation on the left and the page TOC on the right (I've attached some screenshots we're I've redacted any sensitive information)

<img width="2032" alt="screen shot 2018-04-23 at 22 52 49" src="https://user-images.githubusercontent.com/83561/39152780-a7da2e66-4749-11e8-991c-ee3db5fd4ed5.png">
<img width="2032" alt="screen shot 2018-04-23 at 22 53 05" src="https://user-images.githubusercontent.com/83561/39152781-a7f81a7a-4749-11e8-804e-71f1013f22b2.png">

gabro commented:
@mads-hartmann that was actually Hugo's material theme, but yep, it looked very similar.

That said, this morning I started from scratch (yes, again!) and ended up with http://scalameta.org/metals/.

I'm now using [Docusaurus](https://docusaurus.io/docs/en/search.html), which I've found to be quite amazing:

- the default theme just works
- splashy home page + docs + blog + static pages 
- search (with algolia)
- "edit page" button (baked in)
- markdown for content
- React for other pages (I may be biased, but React is 💯)

In one morning I achieved more than I was able two after two days of hacking on Hugo, Gitbook or similar, so I'm very happy with it!

If we're happy with the current implementation, I would suggest we decide on a basic content structure, merge this to test the publishing, fix the publishing (yes, it will happen), and start iterating on the content.

Wdyall think?

mads-hartmann commented:
I think this looks great @gabro 👍 

Do I understand correctly that using this setup it won't be possible to execute Scala code as part of the build? Scalafmt's site currently does so to make sure that the formatting examples on the site contains the actual formatting as performed by scalafmt.

gabro commented:
@mads-hartmann great point. I removed the sbt integration on purpose in this PR, to keep this simple. That said, I think it would be possible to add a pre-processing step (maybe using `sbt-site`) which runs Scala code using `tut:passthrough`.

My idea would be to pre-process the `docs` folder and to feed it to docusaurus later in the publishing pipeline. This should also allow to insert dynamic variables from the build (such as versions and others)

olafurpg commented:
This is amazing @gabro! 🤗

Overall I think this is a great foundation, splashy landing page, readable structure, good mobile view, edit on github, so many check boxes!

I tried docusaurus last December and it's become more mature since then.

Re executing Scala code, it's possible to accomplish that with tut, especially thanks to the passthrough filter. It will require some work to migrate from scalatex but the benefit of standard markdown is worth it.

Re structure of content, I'd like to take a stab at making the overview fit better into the website. Will return with comments or commits later :)

Re react for custom pages, does that mean we could have a playground for scalafmt? 😁

Once the site is live I can contact algolia to get search up.

gabro commented:
> Re react for custom pages, does that mean we could have a playground for scalafmt? 😁

😇 yes!

> Once the site is live I can contact algolia to get search up.

Ah, I've already done so :)

mads-hartmann commented:
@gabro That sounds awesome 🎉 Great job

gabro commented:
I've added the pre-processing step, which allows to:

- inject variables from the build (example in https://github.com/scalameta/metals/pull/294/commits/2b26be14b2aa785989d73fd5af6bfe87cedd689f)
- evaluate `tut` code blocks (example in https://github.com/scalameta/metals/pull/294/commits/4d9bcff18865d820bf0fe47175035aadf67523ea)

The steps to build the website are now:

```
sbt website/preprocess:preprocess
cd website
yarn start
```

/cc @mads-hartmann

gabro commented:
We also have search, now! 

The only bad news, is that as soon as you start typing, Chrome marks the site as Not Secure with a red warning sign. @olafurpg how about we use a free Cloudflare account to serve scalameta.org over https?

mads-hartmann commented:
@gabro Fantastic! Can't wait to have this for the scalafmt site as well ;)

olafurpg commented:
Great idea to enable https! What is involved in setting up cloudflare? We can probably redirect the dns for the domain, cc/ @xeno-by 

Excited as well to use this for scalafmt :) it'll require a bit of work however

gabro commented:
Cloudflare is super easy to setup: dns redirect and a couple of clicks

olafurpg commented:
I'm on https 💪

olafurpg commented on project/travis-deploy-key:
```diff
@@ -0,0 +1,51 @@
+-----BEGIN RSA PRIVATE KEY-----
+MIIJKgIBAAKCAgEAuNPCudT6bimCvdBm/3z6kbPC2ZoBJH52JoLEWK0MmHR2Owcr
```
Delete? 😅

gabro replied:
Woooooops :O delete and change

olafurpg commented on website/blog/2017-04-10-blog-post-two.md:
```diff
@@ -0,0 +1,18 @@
+---
+title: New Blog Post
```
Delete?

olafurpg commented on website/pages/en/index.js:
```diff
@@ -0,0 +1,231 @@
+ * LICENSE file in the root directory of this source tree.
+ */
+
+const React = require('react');
```
![wow](https://user-images.githubusercontent.com/1408093/39279566-776d8d30-48e9-11e8-94db-152429c11b1c.jpg)

olafurpg commented on website/blog/2016-03-11-blog-post.md:
```diff
@@ -0,0 +1,18 @@
+authorFBID: 100002976521003
+---
+
+Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque elementum dignissim ultricies. Fusce rhoncus ipsum tempor eros aliquam consequat. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vivamus elementum massa eget nulla aliquet sagittis. Proin odio tortor, vulputate ut odio in, ultrices ultricies augue. Cras ornare ultrices lorem malesuada iaculis. Etiam sit amet libero te ...
```
Remove?

olafurpg commented on docs/new-editor.md:
````diff
@@ -101,12 +104,24 @@ client is expected to send a `workspace/didChangeConfiguration` notification
+  import scala.meta.metals.Configuration
+  import io.circe.syntax._
+  println("```json")
+  println(Configuration().asJson.spaces2)
````
tut:passthrough 🙏 it's so good

As a general recommendation, I think it's best to keep the code as much as possible in `*.scala` files and call from here. In this case we could implement this body in Configuration.scala to minimize the code in this markdown file. Alternatively, a `object Site { def defaults: String = ... }` that we could call with `Site.default`

olafurpg commented on website/pages/en/index.js:
```diff
@@ -0,0 +1,231 @@
+  return siteConfig.baseUrl + (language ? language + '/' : '') + page;
+}
+
+class Button extends React.Component {
```
how much work would it be to reuse the same setup for scalafmt/scalameta docs? Would we have to copy-paste these components or is there a smarter way to share code?

gabro replied:
Docusaurus provides some basic building blocks (https://docusaurus.io/docs/en/api-pages.html#provided-components), the rest we'll have to copy paste.

olafurpg commented:
Looks like our lucky day, GitHub just added https support for custom domains via letsencrypt https://blog.github.com/2018-05-01-github-pages-custom-domains-https/ we need to update the A record for scalameta.org, but we shouldn't need to configure cloudfront anymore

gabro commented:
Nice! They even offer some of the cloudflare features like CDN and anti-DDOS

olafurpg commented:
https should be enabled now, my laptop hasn't picked it up yet but my phone has

gabro commented:
I've cleaned up a few things, rebased and added a contribution guide for the website itself.

@olafurpg please take a look once you have a minute and I think we can merge it if you don't have further comments

laughedelic commented:
A minor thing: could you switch the "Docs" and "GitHub" links in the toolbar to match the buttons "View on Github" and "Documentation"? This twist is quite noticeable in the mobile version

gabro commented:
Ok, I think I've addressed all the points.
About CONTRIBUTING.md, I've included its content into a `Getting started` page in the docs, under the contributors section. I think it's nicer to have all the info in one place. I've left the file with a pointer to the website, so that potential contributors landing on GitHub will easily find all the information.

I've also upgraded docusaurus, which makes the whole theme a bit "lighter" (sidebar background is transparent, code blocks are more subtle) and it fixes the issue with the alignment of anchors on section headers.

olafurpg commented on CONTRIBUTING.md:
```diff
@@ -1,143 +1,3 @@
-* If you encounter `Error: Channel has been closed` in VSCode console and nothing seems to be working, open Settings
-(<kbd>CMD</kbd> + <kbd>,</kbd> on macOS)
-and make sure the `"metals.serverVersion"` setting points to full version string of the server you have installed, or another existing version of the server (either locally or remotely published).
+Please refer to the documentation for contributors [on our website](https://scalameta.org/metals/docs/getting-started-contributors.html#troubleshooting).
```
👍

olafurpg commented on docs/installation-contributors.md:
````diff
@@ -12,12 +17,10 @@ These steps are required once per machine.
 
 ```scala
-addSbtPlugin("org.scalameta" % "sbt-metals" % "<version>")
+addSbtPlugin("org.scalameta" % "sbt-metals" % "@VERSION@")
````
The "N.B" and "troubleshooting tips" links on this page give a 404

```md
***N.B.***  This project is still in development - right now you will need to run the VSCode plugin as described in the [contributing guide](../CONTRIBUTING.md#running-a-local-version-of-the-vscode-extension)

...

[**Troubleshooting tips**](../CONTRIBUTING.md#troubleshooting)
```

Seems we can remove them.

gabro commented:
Ok, I've moved the Troubleshooting page on its own (I believe it's easier to spot this way) and fixed the broken links. I'm going to go ahead, merge, and see whether the CI works.
